clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2763,14 +2763,18 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
   }
 
   static VarDecl *createKernelObjClone(ASTContext &Ctx, DeclContext *DC,
-                                       const CXXRecordDecl *KernelObj) {
+                                       const CXXRecordDecl *KernelObj,
+                                       FunctionDecl *KernelCallerFunc) {
     TypeSourceInfo *TSInfo =
         KernelObj->isLambda() ? KernelObj->getLambdaTypeInfo() : nullptr;
+    auto Type = QualType(KernelObj->getTypeForDecl(), 0);
+    Type->getAsRecordDecl()->setAnonymousStructOrUnion(true);
```

Comment:
@intel/dpcpp-cfe-reviewers, I'm not sure I'm doing everything right here. 
If `setAnonymousStructOrUnion` is not called, CodeGen library fails at mangling type.
KernelObj type is a lambda expression, so setting anonymous struct sounds okay. Right?

Reply:
Why is it starting to fail now? We never had to do this earlier.

Reply:
According to my understanding it's because this change changes declaration scope of LLVM symbol from function to a module.
Compiler emits SPIR kernel function like this:
```c++
void spir_kernel_function(/*args*/) {
  lambda_type var;
  ...
}
```
This patch adds work-group scope attribute to `var` declaration, which effectively equivalent to adding `static` keyword (or OpenCL `__local` if want to be more precise).
```c++
void spir_kernel_function(/*args*/) {
  __local lambda_type var;
  ...
}
```
LLVM codegen emits a global variable for `var` and to avoid naming conflicts with other globally visible variables with the same name it mangles declaration scope.

I think we haven't seen the failure before because `var` was always a function scope variable and codegen do not mangle names for those.

Reply:
I see. I'm not very sure about this change. What happens if kernel object is a functor and is declared in multiple kernel functions. The same TU can have multiple kernels right? I'm also not sure what the implications of marking something which isn't actually an `anonymousStructOrUnion `as an `anonymousStructOrUnion `is. @premanandrao @erichkeane can you please comment here?

Reply:
Thanks for putting it into words @erichkeane.  I wasn't sure about the change either but couldn't really express why. :-)

Reply:
My concern is - if this variable actually does have file scope, shouldn't we be mangling it ?

Reply:
> What happens if kernel object is a functor and is declared in multiple kernel functions. The same TU can have multiple kernels right?

Right, the same TU can have multiple kernels, but in case of a functor there must be different factors. SYCL compiler emits a single kernel function for all functor invocations because it uses a functor type to produce a kernel function.
But I think you are right that we should check if kernel is a lambda here.

> Marking a lambda 'static' would definitely keep it still mangled, so it seems that the patch is likely doing OTHER things wrong to cause the mangling issue.

I don't think that this patch is doing OTHER things wrong per se. Most likely there is something wrong with the way we build kernel function object variable today or not doing in this patch to make this variable "static". @erichkeane, do you know how FE builds AST for static lambda variables?
The issue can be reproduced if instead of adding work-group scope attribute we change `SC_None` with `SC_Static`. The attribute does basically the same + allocates the variable in `__local` address space. We hit this assert: https://github.com/intel/llvm/blob/sycl/clang/lib/AST/ItaniumMangle.cpp#L1516.

> My understanding is marking it as `anonymousStructOrUnion `is avoiding mangling, making it behave like it used to?

Is it? I'm not sure, that's why brought your attention to this function. Isn't it mangled the same way regular lambda expression type is mangled?

Reply:
From the looks of this: https://godbolt.org/z/ha9KE7Khb 
'static' doesn't seem to change the mangling, so it likely only modifies the linkage.

Reply:
I've applied "anonymous structure" flag to lambda types only.